prometheus/benchmark_test.go

```diff
@@ -43,17 +45,39 @@ func BenchmarkCounterWithLabelValuesConcurrent(b *testing.B) {
 	)
 	b.ReportAllocs()
 	b.ResetTimer()
-	wg := sync.WaitGroup{}
-	for i := 0; i < 10; i++ {
-		wg.Add(1)
-		go func() {
-			for j := 0; j < b.N/10; j++ {
-				m.WithLabelValues("eins", "zwei", "drei").Inc()
-			}
-			wg.Done()
-		}()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			m.WithLabelValues("eins", "zwei", "drei").Inc()
+		}
+	})
+}
+
+func BenchmarkCounterWithCardinalitiesParallel(b *testing.B) {
+	labels := []string{"one", "two", "three", "four", "five", "six", "seven"}
+	values := []string{"eins", "zwei", "drei", "vier", "fünf", "sechs", "sieben"}
+	for i := 1; i <= len(labels); i++ {
+		b.Run(fmt.Sprintf("%d vectors", int(math.Pow(float64(i), float64(i)))), func(b *testing.B) {
+			m := NewCounterVec(
+				CounterOpts{
+					Name: "benchmark_counter",
+					Help: "A counter to benchmark it.",
+				},
+				labels[:i],
+			)
+			b.ReportAllocs()
+			b.ResetTimer()
+			b.RunParallel(func(pb *testing.PB) {
+				r := rand.New(rand.NewSource(42))
+				for pb.Next() {
+					vs := make([]string, i)
+					for j := 0; j < i; j++ {
+						vs[j] = values[r.Intn(i)]
+					}
+					m.WithLabelValues(vs...).Inc()
+				}
+			})
```

Comment:
I am interested in the "warmup" itself when cardinality growth, that is the reason I use an exponential-growthed cardinality variable to observe it, this benchmark test makes it easy to see all cases, from all measurements hit the cache (1 ^ 1) to all misses the cache (7 ^ 7). Because Prometheus client uses a `sync.Map` liked structure to index cached ekements, I think if all series has already "warmuped", it should be an O(1) time complexity without cardinality sensitive.

Or I can add another test and "warmup" before the test to as a control group to previous test, what is your suggestion?

Reply:
That all makes sense.

I guess we can safely assume that the O(1) assumption holds in the steady-state case. So I wouldn't be keen on benchmarking that. (But if anyone is, I wouldn't be against it.)

To really test the "warmup" itself, I would say we should fill the vector systematically (not randomly) with all possible combinations, and one of those complete fills should be one iteration of the benchmark. (The one for 7 ^ 7 will take a while, but the Go benchmarking can handle that. It will just run very few iterations.) In that way, each iteration for a given label count is of the same length, which is what the Go benchmarking expects.

To compare the results, you then have to normalize the runtime per iteration "manually", i.e. in the ideal case, you would expect the runtime for the 7 ^ 7 case to be 823543 times as long as for 1 ^ 1, but it will in practice be more than that, which tells us the performance less because of cache misses etc.

Alternatively, you could make the 1 ^ 1 iteration to fill the one element into the vector 823543 times, then the 2 ^ 2 case to fill the 4 elements 205885 times, etc., until finally filling the 823543 elements into the vector for the 7 ^ 7 case exactly once.

I can try to explain it better, but perhaps you can follow me already. Please let me know.

Reply:
I think I got your advice, I will try to iterate all permutations from 1 ^ 1 to 7 ^ 7 rather than get it randomly. If I get something wrong, please tell me.

Reply:
I think maybe we do not need to make all rounds have the same iteration number, just use `for pb.Next() {}`  and let Go Bench handle this by itself.

Reply:
Just give it a try, and we can see how it works.